### PR TITLE
[SPARK-35032][PYTHON] Port Koalas Index unit tests into PySpark

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Now that we merged the Koalas main code into the PySpark code base (#32036), we should port the Koalas Index unit tests to PySpark.

### Why are the changes needed?
Currently, the pandas-on-Spark modules are not tested fully. We should enable the Index unit tests.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Enable Index unit tests.

Keyword:SPARK-34849

### CODE REVIEW

File: python/pyspark/pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -0,0 +1,2308 @@
...
+            kmidx = ps.from_pandas(pmidx)
+            self.assert_eq(kmidx.is_monotonic_increasing, False)
+            self.assert_eq(kmidx.is_monotonic_decreasing, False)
+
+            pmidx = pd.MultiIndex.from_tuples([(None, None)])
+            kmidx = ps.from_pandas(pmidx)
+            self.assert_eq(kmidx.is_monotonic_increasing, False)
+            self.assert_eq(kmidx.is_monotonic_decreasing, False)
+        else:
+            # Disable the test cases below because pandas returns `True` or `False` randomly.
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/tests/indexes/test_base.py:1256]: Please refer to https://github.com/apache/spark/runs/2329674423 for the test failure.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Comment by xinrong-databricks** [comment]
cc @HyukjinKwon  @itholic

**Comment by itholic** [comment]
Nice work. LGTM

**Comment by HyukjinKwon** [comment]
Hm, one test failed:
```
======================================================================
2021-04-15T03:54:15.0802269Z ERROR [61.702s]: test_monotonic (pyspark.pandas.tests.indexes.test_base.IndexesTest)
2021-04-15T03:54:15.0803808Z ----------------------------------------------------------------------
2021-04-15T03:54:15.0804406Z Traceback (most recent call last):
2021-04-15T03:54:15.0835123Z   File "/__w/spark/spark/python/pyspark/pandas/tests/indexes/test_base.py", line 1280, in test_monotonic
2021-04-15T03:54:15.0836254Z     self.assert_eq(kmidx.is_monotonic_increasing, pmidx.is_monotonic_increasing)
2021-04-15T03:54:15.0837366Z   File "/__w/spark/spark/python/pyspark/pandas/testing/utils.py", line 293, in assert_eq
2021-04-15T03:54:15.0838145Z     self.assertEqual(lobj, robj)
2021-04-15T03:54:15.0838748Z AssertionError: False != True
2021-04-15T03:54:15.0839117Z 
2021-04-15T03:54:15.0840085Z ----------------------------------------------------------------------
2021-04-15T03:54:15.0840665Z Ran 76 tests in 290.024s
2021-04-15T03:54:15.0840965Z 
2021-04-15T03:54:15.0841355Z FAILED (errors=1)
```

File: dev/sparktestsupport/modules.py [hunk 2]
```diff
@@ -611,43 +611,47 @@ def __hash__(self):
         "pyspark.pandas.spark.utils",
         "pyspark.pandas.typedef.typehints",
         # unittests
-        "pyspark.pandas.tests.test_dataframe",
+        "pyspark.pandas.tests.indexes.test_base",
```
Comment (Yikun) [inline dev/sparktestsupport/modules.py:614]: Looks like we could add some testcase discover mechanism on it in follow up.

File: python/pyspark/pandas/tests/indexes/test_datetime.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+            self.assert_eq(
+                kidx.indexer_between_time("00:00:00", "00:01:00", False, False).sort_values(),
+                pd.Index(pidx.indexer_between_time("00:00:00", "00:01:00", False, False)),
+            )
+
+            self.assert_eq(
+                kidx.indexer_between_time("00:00:00", "00:01:00", False, False).sort_values(),
+                pd.Index(pidx.indexer_between_time("00:00:00", "00:01:00", False, False)),
+            )
```
Comment (Yikun) [inline python/pyspark/pandas/tests/indexes/test_datetime.py:167]: redundant call, looks like you want a True True case?

**Review by Yikun — CHANGES_REQUESTED** [review]
LGTM, except a nit

File: python/pyspark/pandas/tests/indexes/test_datetime.py (hunk 3 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/indexes/test_datetime.py]: It's updated. Good catch!

File: dev/sparktestsupport/modules.py (hunk 2 above)
Comment (xinrong-databricks) [reply dev/sparktestsupport/modules.py:614]: Sounds good! That would reduce maintenance costs.

**Comment by xinrong-databricks** [comment]
@HyukjinKwon Good catch! pandas returns `True` or `False` randomly for `test_monotonic`. A few test cases are disabled now.

**Review by ueshin — APPROVED** [review]
The changes since I took a look las time LGTM, pending tests.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: dev/sparktestsupport/modules.py (hunk 2 above)
Comment (Yikun) [reply dev/sparktestsupport/modules.py:614]: Finally, it is merged in https://github.com/apache/spark/pull/32867 , : )

Comment (xinrong-databricks) [reply dev/sparktestsupport/modules.py:614]: Thanks for making it happen! :)

**Pull request closed without merging**